### PR TITLE
Implementation sketch

### PR DESCRIPTION
**Pull request by gabi-250**
This PR contains the first draft of the implementation:
* e2d203e `Config` (the configuration struct used to configure experiments) now also contains `results_dir` (the path of the directory where `k2` stores all the files it creates).
* c579fc1 adds the manifest manager. I know `krun` has its own manifest format, but I wasn't sure we wanted this for `k2` as well. The `k2` manifest is a json file with all the information needed  to resume execution after rebooting (it's still missing a couple of values we will need, like the starting temperatures). For example, the manifest might look like this:
```
{
    "num_reboots": 6,
    "num_mails_sent": 0,
    "curr_job_idx": 4,
    "jobs": [
        {
            "key": "/usr/bin/python:./benchmarks/binarytrees/binarytrees.py",
            "bench_idx": 0,
            "status": "Done"
        },
        { ... },
        ...
    ]
}
```

[I don't think we actually need the `key` field, since I'm using `bench_idx` as an identifier for benchmarks]

Let me know if you think another format would work better.

* d6d2416 contains the code that actually runs the benchmarks. The experiment runner runs the next outstanding benchmark, and [retrieves its tags](https://github.com/softdevteam/k2/blob/d6d24167bff4656d87652014b01fee6f0c86534d/src/experiment.rs#L60). The tags are supposed to contain all the values measured/recorded, but I haven't implemented that yet, so the 'measurements' are discarded (they'll eventually be stored in an SQLite database). After each benchmark run, the manifest is updated and [explictly written to disk](https://github.com/softdevteam/k2/blob/d6d24167bff4656d87652014b01fee6f0c86534d/src/experiment.rs#L72-L75).
* cd9cffb adds a `reboot` function. If `hardware_reboots` is `false`, this function simulates a reboot (like `krun`) by replacing the process image with a fresh version of itself. `reboot` is unimplemented for the `hardware_reboots == true`case .

* 357e84b adds the `binarytrees` benchmark used in `example/simple_example.rs`

### CODE REVIEW

**Comment by vext01** [comment]
> c579fc1 adds the manifest manager. I know krun has its own manifest format, but I wasn't sure we wanted this for k2 as well.

Ah, we should have discussed the manifest format earlier.

What we learned from Krun is that the manifest file has to be designed in such a way that loading the whole file into memory is never necessary. We do this because we want to minimise non-benchmarking-related memory consumption (and thus memory fragmentation too).

The problem with JSON is that most parsers require the whole file to be loaded in at once. You can get streaming parsers, but they are few and far between.

The Krun format uses a fixed-width record format, and the next job is found by scanning the file line by line until an `O` (outstanding) record is found. The line-wise approach prevents the need to load the whole file at once. And because each record is fixed-size, it means you can update it in-place without skewing the offsets of other records.

You will notice that in Krun manifest files, some numeric fields are padded with zeroes, e.g.:
```
num_mails_sent=0000
num_reboots=00000400
```

That's because these fields are mutable, but we don't want an update to make the lines longer/shorter. Thus we "overallocate" enough space (leading zeroes).

We might be able to improve upon the manifest format, but we probably do want to keep the memory properties intact if we can.

I'm open to some out of the box thinking here, as dealing with the Krun manifest is a bit awkward and low-level (as I'm sure you've seen). Can you think of any better way?

File: benchmarks/binarytrees/binarytrees.lua [hunk 1]
```diff
@@ -0,0 +1,48 @@
+-- https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
```
Comment (vext01) [inline benchmarks/binarytrees/binarytrees.lua:32]: What's the license of this file? Is it compatible with our license?

**Comment by gabi-250** [comment]
@vext01 Aha, that makes perfect sense! I'll see if I can come up with any improvements to the original Krun manifest.

**Comment by ltratt** [comment]
I have a possibly cunning suggestion for the manifest format, though it will take a bit of investigating. If we use Serde + https://crates.io/crates/bincode then I think we might get the properties we want without too much fuss (and also have a very small on disk representation). I had a vague memory that bincode involved zero copying, though the current docs don't mention that, so perhaps I dreamed that... Anyway, it might be a good idea, it might be a bad idea: take it for what it's worth :)

File: src/experiment.rs [hunk 2]
```diff
@@ -38,20 +94,20 @@ impl<'a> Experiment<'a> {
...
 /// to run.
 pub struct ExperimentBuilder<'a> {
-    config: Config,
+    config: Config<'a>,
     benchmarks: Vec<&'a Benchmark<'a>>,
 }
 
 impl<'a> ExperimentBuilder<'a> {
-    pub fn new() -> Self {
+    pub fn new(results_dir: &'a str) -> Self {
```
Comment (vext01) [inline src/experiment.rs]: Should this accept any type which can be converted to a `Path`?

**Comment by vext01** [comment]
> If we use Serde

I did wonder about serde. If it doesn't support zero-copy, you can still encode multiple  data structures in sequence (as `ykpack` does), thus emulating the line-by-line approach of the Krun manifest format (albeit struct-by-struct).

If you choose a fixed-width encoder, and make every record a variant of the same enum, then the size of each record should also be constant. I guess most encoders use variable-size encodings for efficiency though...

**Comment by ltratt** [comment]
> I guess most encoders use variable-size encodings for efficiency though...

For compound types, almost certainly: for things like ints, and so on, I'm not so sure. It's probably worth a quick check.

**Comment by vext01** [comment]
Looks like msgpack supports zero-copy:

> RMP allows to decode bytes from a buffer in a zero-copy manner easily and blazingly fast, while Rust static checks guarantees that the data will be valid as long as the buffer lives.

https://github.com/3Hren/msgpack-rust

**Comment by ltratt** [comment]
Sounds like it might be ideal!

**Comment by vext01** [comment]
Worth a shot!

The only downside is that binary formats are not human readable, however, I think we can probably live with that.

**Comment by vext01** [comment]
So with zero-copy, would you be able to decode individual elements of a vector without reading the whole vector?

**Comment by ltratt** [comment]
I would imagine that it would depend on the type of the things in the vector. A `Vec<u64>` is going to be pretty easy, for example.

**Comment by vext01** [comment]
We should check, but I suspect you will still have to load the whole data structure in at once.

So maybe we might need the `ykpack` trick anyway...

**Comment by ltratt** [comment]
Yes, but don't forget that we don't need to solve the general case: we only need things we want in the manifest to be serialised appropriately. And that's probably very simple stuff.

**Comment by vext01** [comment]
>  we only need things we want in the manifest to be serialised appropriately

I understand, but we ideally don't want to use more memory than necessary.

The experiment we just finished over the weekend ran 400 process executions, each of which has a manifest record. If you can't read each record individually then you have to allocate space for 400 records in memory.

A better strategy is one where you scan one record at a time (thus requiring space for only one in-memory record) until you find the one you want.

The ykpack way would be to have a record type and then you serialise 400 of them independently.

**Comment by vext01** [comment]
Having thought about this a little more, I think we can do even better. If you look at a Krun manifest file, it looks like this:

```
eta_avail_idx=40
num_mails_sent=0000
num_reboots=00000400
keys
O hotspot:binarytrees
O hotspot:fannkuch
O pypy:binarytrees
O pypy:fannkuch
O hotspot:binarytrees
O hotspot:fannkuch
O pypy:binarytrees
O pypy:fannkuch
...
```

(not an actual manifest file (elided), but you get the idea)

It consists of a header (everything up until `keys`), and then a repeating sequence of benchmark keys and their status. Here we have repeated:
```
O hotspot:binarytrees
O hotspot:fannkuch
O pypy:binarytrees
O pypy:fannkuch
```

So here's my idea. If we ensure the order of the repeating keys is *deterministic*, then it suffices to store only the index of the next key. The runner doesn't care about the statuses of past benchmarks to know what to run next, so that can be stored elsewhere (i.e. sqlite).

So I think we can store a simple statically sized header like this:
```
eta_avail_idx=40
num_mails_sent=0000
num_reboots=00000400
next_idx=6
```

So `next_idx` is a new field and you can use modular arithmetic to figure out which key that corresponds to. Here 6 % 4 = 2, which corresponds with `pypy:binarytrees`.

Is that any better?

**Comment by ltratt** [comment]
We still have to load the benchmark names (or paths or whatever) back in, but, yes, I think something simple like that could help simplify the other aspects.

**Comment by vext01** [comment]
The only case where that strategy becomes fiddly is when we have differing numbers of process executions for different keys. We may choose to block the user from doing that?

**Comment by gabi-250** [comment]
Thank you for the comments!

> So here's my idea. If we ensure the order of the repeating keys is deterministic, then it suffices to store only the index of the next key. The runner doesn't care about the statuses of past benchmarks to know what to run next, so that can be stored elsewhere (i.e. sqlite).

I agree, a similar idea occurred to me as I was reading the comments.

> We still have to load the benchmark names (or paths or whatever) back in, but, yes, I think something simple like that could help simplify the other aspects.

I don't think the benchmark paths have to be loaded at all, right? An `Experiment` contains the [list of benchmarks to run](https://github.com/softdevteam/k2/blob/2600741a739f476d441ad250f2641dde63822686/examples/simple_experiment.rs#L40), so we can just use `next_idx` to retrieve the right one.

> The only case where that strategy becomes fiddly is when we have differing numbers of process executions for different keys. We may choose to block the user from doing that?

In addition to `next_idx`, we could also record `execs_left` (the number of outstanding executions for the benchmark identified by `next_idx`). We would have to run each benchmark `pexec` times before moving on to the next one, so we'd only increment `next_idx` when `execs_left` becomes `0` (the value of `pexec` can be different for each benchmark). Here, `next_idx` is the key of the next benchmark to run. For example, we might run the benchmarks in this order:
 
```
O hotspot:binarytrees
O hotspot:binarytrees
O hotspot:binarytrees
O hotspot:fannkuch
O hotspot:fannkuch
O pypy:binarytrees
O pypy:fannkuch
O pypy:fannkuch
```

The manifest below states that `hotspot:fannkuch` (with index 1) has to run one more time before `pypy:binarytrees` can run.

```
eta_avail_idx=40
num_mails_sent=0000
num_reboots=00000400
next_idx=1
execs_left=1
```

Do you think this would work?

**Comment by vext01** [comment]
> For example, we might run the benchmarks in this order:

I prefer to "interlace" the benchmarks. Its arguably better for benchmarking quality, but also it's useful to investigate the data so far before an experiment has completed. By interlacing the benchmarks you get a more even spread early on.

> I don't think the benchmark paths have to be loaded at all, right? An Experiment contains the list of benchmarks to run

I think Laurie means the benchmark instances themselves (which contain the names etc) have to be loaded, but I can't see any avoiding that TBH (unless you try to load them lazily somehow, but now it's getting silly ;) )

**Comment by gabi-250** [comment]
> I prefer to "interlace" the benchmarks. Its arguably better for benchmarking quality, but also it's useful to investigate the data so far before an experiment has completed. By interlacing the benchmarks you get a more even spread early on.

I see. Why would interlacing them improve the quality though? Wouldn't you reboot between proess executions anyway?

**Comment by vext01** [comment]
> Why would interlacing them improve the quality though?

Well, I'm not even totally sure, so let me know your opinion:

Suppose we have two benchmarks and they have equal runtime and we run them twice each. Now suppose *something* happens which adds noise to the first half of the benchmarking session.

Is it better for the noise to be distributed between the two benchmarks, or all concentrated on one?

[It's a flaky argument I know]

**Comment by gabi-250** [comment]
@vext01 You make an interesting point. Wouldn't most 'anomalies' like that be caught by temperature/dmesg checks though? I guess it would be safer to interlace jobs anyway, even if that means  restricting users to having the same number of process executions for all jobs. Still, if a job runs far slower than all others, this constraint might be a problem. 

Anyway, maybe it's too early to worry about this. I'll implement the version you proposed.

**Comment by ltratt** [comment]
I tend to prefer running jobs A, B, & C in some non-consecutive order (either random or sequenced like ABCABCABC) in case there's some aspect we haven't thought of that is time based (e.g. my personal favourite, cosmic radiation, effecting a box more than normal for a certain time period: it's better that it affects all benchmarks a bit rather just affecting, say, A's pexecs).

**Comment by vext01** [comment]
> either random

Random ordering may complicate the indexing strategy above, but if we want it, we could use a fixed random seed (or a random seed stored in the manifest)

**Comment by gabi-250** [comment]
> I tend to prefer running jobs A, B, & C in some non-consecutive order (either random or sequenced like ABCABCABC) in case there's some aspect we haven't thought of that is time based (e.g. my personal favourite, cosmic radiation, effecting a box more than normal for a certain time period: it's better that it affects all benchmarks a bit rather just affecting, say, A's pexecs).

That makes sense, although I must admit I've never really given much thought to these kinds of external factors before!

> Random ordering may complicate the indexing strategy above, but if we want it, we could use a fixed random seed (or a random seed stored in the manifest)

I don't mind implementing either of the two orderings, although I'm still a bit unsure whether a random ordering would be better than the ordering used by Krun. I don't feel strongly about it though.

**Comment by ltratt** [comment]
Personally I have a mild preference for fully random ordering, but I'm also happy with ABCABCABC. @vext01 I might have misunderstood you, but I assumed that we set the ordering at the point we create the manifest file, write out the benchmarks that we'll run (in whatever order we decide at manifest creation time), and then we can just use your "index" trick to iterate through them?

**Comment by vext01** [comment]
> I assumed that we set the ordering at the point we create the manifest file... write out the benchmarks that we'll run (in whatever order we decide at manifest creation time)

The [idea I proposed earlier]( https://github.com/softdevteam/k2/pull/3#issuecomment-522607499) was to try to avoid the need to write out the sequence in full by exploiting the predictability of the repeating sub-sequence.

Randomisation make that strategy harder, however, even *with* randomisation, [as alluded to earlier](https://github.com/softdevteam/k2/pull/3#issuecomment-523107204) our random sequence of benchmarks could be reconstructed from a random seed (stored in the manifest). It'd add complexity, but I think is doable. Then again, we might be going overboard...

**Comment by ltratt** [comment]
> The idea I proposed earlier was to try to avoid the need to write out the sequence in full by exploiting the predictability of the repeating sub-sequence.

Oh, yeah, I see what you mean now. Sorry, I didn't appreciate that before!

I can imagine (slightly contrived) situations where we decide that we really do need to control the order in advance, so I'm inclined to suggest that the manifest file should contain the order we want in advance, rather than reconstructing it on the fly. It gives us more flexibility: if we make the record size fixed, then we can still use your "just store the index" trick I think?

**Comment by vext01** [comment]
> if we make the record size fixed, then we can still use your "just store the index" trick I think?

Yep, the manifest file can just be a list of fixed-size integer indices. Then you'd just need a k2 to maintain a deterministic list that these indices correspond to.

With some skill you might be able to avoid allocating the list of benchmarks in memory too, but I've not fully fleshed that idea out in my head. For now, let's just allocate it, and later we can decide if we want to optimise. At least the list will be constant size between reboots, so it's not a disaster.

On the other hand, if the manifest file contains the list itself, then you wouldn't need to allocate the list in-memory! Gah!

Not sure what's best now...

**Comment by gabi-250** [comment]
I ended up implementing the fully random ordering.

The manifest now consists of a header, stored in a text file called `manifest.k2`, and a body. The format of the header is still incomplete, since k2 doesn't record the starting temperatures yet. For example, `manifest.k2` might look like this:

```
num_mails_sent=0000
num_reboots=00000006
next_idx=0003
ordering=4,5,2,0,3,1
```

The `ordering` record indicates the order in which jobs will be scheduled. Since it's the last record, its width can vary without affecting the offsets of the other records. Although the other records are fixed-width, the current implementation doesn't update them in-place. Instead, the entire header is written to disk after every job run: https://github.com/softdevteam/k2/blob/0a7a2cc0129f629ff15ce0d72020c2798b96d1bc/src/manifest.rs#L178-L187

In my next PR, I'm going to change `sync_header` to use the offsets of the manifest records to only rewrite the mutable part of each line.

The [parse](https://github.com/softdevteam/k2/blob/0a7a2cc0129f629ff15ce0d72020c2798b96d1bc/src/manifest.rs#L64-L107) function of the `ManifestHeader` is rather ugly, but I can't think of any other way to write it.


The manifest body is stored in the SQLite database in the `manifest` [table](https://github.com/softdevteam/k2/blob/0a7a2cc0129f629ff15ce0d72020c2798b96d1bc/src/db.rs#L59-L72). It consists of [Jobs](https://github.com/softdevteam/k2/pull/3/files#diff-029ddd8a7a967b59c991c711bec806ccR30) serialised as JSON. However, I'm not totally sure there's any advantage to using JSON here...

The database is only accessed after running the current job, so it shouldn't affect the results: https://github.com/softdevteam/k2/pull/3/commits/0a7a2cc0129f629ff15ce0d72020c2798b96d1bc#diff-2a79734ebde24591499a8fba6696f01eR73-R79

(Sorry I've neglected this PR for so long!)

**Comment by gabi-250** [comment]
> What's the license of this file? Is it compatible with our license?

It's a modified 3-clause BSD license, so I believe it is. Just to be sure, I added the license to both benchmarks (91c5244).

**Comment by vext01** [comment]
> ordering=4,5,2,0,3,1

Couldn't you just store a random seed from which the sequence can be reconstructed? Then you can build an iterator over the sequence and use `.skip()` to jump to get the desired job.

> It consists of Jobs serialised as JSON. However, I'm not totally sure there's any advantage to using JSON here...

The structure of this is fixed, right? So it could indeed be a regular relational DB table.

> The database is only accessed after running the current job, so it shouldn't affect the results

So to be clear, to find the next job to run, the (fixed-size) manifest is reconstructed *in-memory* and there is no need to open the database?

**Comment by gabi-250** [comment]
> Couldn't you just store a random seed from which the sequence can be reconstructed? Then you can build an iterator over the sequence and use .skip() to jump to get the desired job.

I could, but Laurie suggested we might actually want to store the ordering in the manifest, as opposed to reconstructing it on the fly. I don't mind either way.

> The structure of this is fixed, right? So it could indeed be a regular relational DB table.

Yep, it is. I'll store it as a regular table then.

> So to be clear, to find the next job to run, the (fixed-size) manifest is reconstructed in-memory and there is no need to open the database?

Yes, that is right. However, the list of `Benchmark` instances is still loaded in memory. The manifest stores the index of the next job, which is used to decide which benchmark to run next: https://github.com/softdevteam/k2/pull/3/files#diff-2a79734ebde24591499a8fba6696f01eR65-R66.

**Comment by vext01** [comment]
> I could, but Laurie suggested we might actually want to store the ordering in the manifest, as opposed to reconstructing it on the fly. I don't mind either way.

Ah, I forgot about that.

It just seems weird to store a massive list in the file, when there's no need. @ltratt how are you feeling about this?

> Yes, that is right. However, the list of Benchmark instances is still loaded in memory. 

I can live with that. I suspect we can make it lazy later anyway, right?

**Comment by gabi-250** [comment]
> I can live with that. I suspect we can make it lazy later anyway, right?
 
Right, that should be doable.

**Comment by ltratt** [comment]
I think generating the list up front is the right thing to do: it means we can randomise it (or not...) as we wish, without worrying about the computational complexity of doing so at "run-time" (so to speak).

**Comment by gabi-250** [comment]
I think I've addressed all your comments (again, sorry for the delay).

@vext01

> That's because these fields are mutable, but we don't want an update to make the lines longer/shorter. Thus we "overallocate" enough space (leading zeroes).
>
> We might be able to improve upon the manifest format, but we probably do want to keep the memory properties intact if we can.

In 92c3390 I changed the `ManifestHeader` to also store the offset of each of its mutable records (currently, these are `num_reboots`, `num_mails_sent` and `next_idx`). The header is now updated like this:

https://github.com/softdevteam/k2/blob/92c339088d6cb9c7ee2b297d2f07bb0aa7d93116/src/manifest.rs#L227-L242

`NUM_MAILS_SENT_BYTES`, `NUM_REBOOTS_BYTES`, and `NEXT_IDX_BYTES` are the widths of the three mutable records. I defined my own format function ([format_int_field](https://github.com/softdevteam/k2/blob/92c339088d6cb9c7ee2b297d2f07bb0aa7d93116/src/manifest.rs#L73-L82)), because the first argument of `format!` must be a string literal, and I figured reusing the `*_BYTES` constants for the formatting would be less error-prone than hard-coding the format string of each record (otherwise, if one of the `*_BYTES` constants changed, you would've had to update the corresponding format string as well).

**Comment by vext01** [comment]
> num_mails_sent

We don't have emailing support yet, do we? I'd be tempted to kill that until we have emailing implemented.

> NUM_MAILS_SENT_BYTES, NUM_REBOOTS_BYTES, and NEXT_IDX_BYTES are the widths of the three mutable records

Do we have overflow checks when incrementing these?

**Comment by ltratt** [comment]
> Do we have overflow checks when incrementing these?

I'm not sure I can see where this could be an issue?

**Comment by vext01** [comment]
> I'm not sure I can see where this could be an issue?

Well, if we allocate (say) 5 ASCII characters for a numeric field in the manifest file, I'd hope that we get a crash with a good error message if the field increments above `99999`, as `100000` would consume 6 ASCII characters.

**Comment by gabi-250** [comment]
> We don't have emailing support yet, do we? I'd be tempted to kill that until we have emailing implemented.

You're right, we don't. I'll remove it.


> > NUM_MAILS_SENT_BYTES, NUM_REBOOTS_BYTES, and NEXT_IDX_BYTES are the widths of the three mutable records

> Do we have overflow checks when incrementing these?

There's a check in the format function:

https://github.com/softdevteam/k2/blob/92c339088d6cb9c7ee2b297d2f07bb0aa7d93116/src/manifest.rs#L79

I'll add similar overflow checks where these records incremented.

File: src/db.rs [hunk 3]
```diff
@@ -37,41 +37,38 @@ impl<'a> K2Store {
...
+            .expect("Failed to create the job table");
         let mut id = 0;
         for _ in 0..config.pexecs {
             for bench in benchmarks {
                 let job = Job::new(id, &bench);
                 id += 1;
+                let query = format!(
+                    "
+                    INSERT INTO job
+                    VALUES ({}, '{}', {});
```
Comment (vext01) [inline src/db.rs]: Shouldn't we be using prepared queries instead of string formatting?

Comment (gabi-250) [reply src/db.rs]: Good point, fixed in 5d762d2.

**Comment by ltratt** [comment]
> Well, if we allocate (say) 5 ASCII characters for a numeric field in the manifest file, I'd hope that we get a crash with a good error message if the field increments above 99999, as 100000 would consume 6 ASCII characters.

Ah, sorry, I had completely misunderstood what you meant!

**Comment by vext01** [comment]
Latest changes LGTM. Are you ready to squash?

**Comment by gabi-250** [comment]
@vext01 Done.

**Comment by vext01** [comment]
By the way, we've moved to doing PRs from personal forks.

**Pull request merged**